crates/swc_ecma_lints/src/rules/dot_notation.rs

```diff
@@ -0,0 +1,189 @@
+use swc_ecma_ast::*;
+use swc_ecma_visit::{noop_visit_type, Visit, VisitWith};
+
+static KEYWORDS: Lazy<AHashSet<&'static str>> = Lazy::new(|| {
+    let mut keywords: AHashSet<&'static str> = AHashSet::default();
+
+    [
+        "abstract",
+        "boolean",
+        "break",
+        "byte",
+        "case",
+        "catch",
+        "char",
+        "class",
+        "const",
+        "continue",
+        "debugger",
+        "default",
+        "delete",
+        "do",
+        "double",
+        "else",
+        "enum",
+        "export",
+        "extends",
+        "false",
+        "final",
+        "finally",
+        "float",
+        "for",
+        "function",
+        "goto",
+        "if",
+        "implements",
+        "import",
+        "in",
+        "instanceof",
+        "int",
+        "interface",
+        "long",
+        "native",
+        "new",
+        "null",
+        "package",
+        "private",
+        "protected",
+        "public",
+        "return",
+        "short",
+        "static",
+        "super",
+        "switch",
+        "synchronized",
+        "this",
+        "throw",
+        "throws",
+        "transient",
+        "true",
+        "try",
+        "typeof",
+        "var",
+        "void",
+        "volatile",
+        "while",
+        "with",
+    ]
+    .iter()
+    .for_each(|keyword| {
+        keywords.insert(*keyword);
+    });
+
+    keywords
+});
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "camelCase")]
+pub struct DotNotationConfig {
+    allow_keywords: Option<bool>,
+    allow_pattern: Option<String>,
+}
+
+pub fn dot_notation(
+    source_map: &Arc<SourceMap>,
+    config: &RuleConfig<DotNotationConfig>,
+) -> Option<Box<dyn Rule>> {
+    match config.get_rule_reaction() {
+        LintRuleReaction::Off => None,
+        _ => Some(visitor_rule(DotNotation::new(source_map.clone(), config))),
+    }
+}
+
+#[derive(Default)]
+struct DotNotation {
+    expected_reaction: LintRuleReaction,
+    allow_keywords: bool,
+    pattern: Option<Regex>,
+    source_map: Arc<SourceMap>,
+}
+
+impl Debug for DotNotation {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("DotNotation")
+            .field("expected_reaction", &self.expected_reaction)
+            .field("allow_keywords", &self.allow_keywords)
+            .field("pattern", &self.pattern)
+            .finish()
+    }
+}
+
+impl DotNotation {
+    fn new(source_map: Arc<SourceMap>, config: &RuleConfig<DotNotationConfig>) -> Self {
+        let dot_notation_config = config.get_rule_config();
+
+        Self {
+            expected_reaction: *config.get_rule_reaction(),
+            allow_keywords: dot_notation_config.allow_keywords.unwrap_or(true),
+            source_map,
+            pattern: dot_notation_config
+                .allow_pattern
+                .as_ref()
+                .map(|patter| Regex::new(patter).unwrap()),
```

Comment:
I think it's better to have a cache for regex.

Reply:
Also, can you improve the error message?
`unwrap()` will fail with a bad error message, so `expect()` with a good error message is preferable.

Reply:
> I think it's better to have a cache for regex.

How I can cache regex in this case? Static mutable hash map?

Reply:
You can refer to https://github.com/swc-project/swc/blob/5e82640b42c11b07973b2849347570b643137018/crates/swc/src/config/mod.rs#L871-L905

Reply:
Thanks